### PR TITLE
VACMS-3324: Fix paragraph and Node being out of sync

### PR DESCRIPTION
**Pull request by indytechcook**
## Description

See #3325.  This fix adds a patch to fix the underlining issue of the paragraph always saving the `latest` revision as the `default` revision.  This was causing issues when the `paragraph` is loaded outside the context of the field.  The specific issue is with CMS export.

Looking at the data from the field we can see the revision paragraph ids.  These are used to test the script and verify the results.  The selection of the node came from the reported issue: https://github.com/department-of-veterans-affairs/va.gov-team/issues/15321

The update script updates the revision on the paragraph tables to the correct revision.  Originally I attempted to re-save the node to update the revision but this solution only worked if there weren't any draft node revisions.  

These queries show the data out of sync between the field storage and paragraph storage

```
select field_content_block_target_id, field_content_block_target_revision_id from node__field_content_block where entity_id=255;
```

| field_content_block_target_id | field_content_block_target_revision_id |
|-------------------------------|----------------------------------------|
| 3784                          | 35659                                  |
| 3805                          | 35660                                  |
| 3807                          | 35661                                  |
| 3809                          | 35662                                  |
| 5510                          | 35663                                  |

```
SELECT id, revision_id
FROM paragraphs_item 
WHERE id in (3784, 3805, 3807, 3809, 5510);
```

| id   | revision_id |
|------|-------------|
| 3784 | 37275       |
| 3805 | 37276       |
| 3807 | 37277       |
| 3809 | 37278       |
| 5510 | 372791      |

## Testing done

Ran the test against the known node which has the issue:

```
$paragraph_ids = [3784, 3805, 3807, 3809, 5510];
$new_nids = runRange($paragraph_ids);
updateParagraphs($new_nids);
```

**Before Data Locally**

```
SELECT t.*
FROM drupal8.paragraphs_item_field_data t
WHERE id in (3784, 3805, 3807, 3809, 5510);
```

| id   | revision_id |
|------|-------------|
| 3784 | 37275       |
| 3805 | 37276       |
| 3807 | 37277       |
| 3809 | 37278       |
| 5510 | 372791      |

**After Data Locally**

| id   | revision_id |
|------|-------------|
| 3784 | 35659       |
| 3805 | 35660       |
| 3807 | 35661       |
| 3809 | 35662       |
| 5510 | 35663       |

The revisions are back in sync with the field data.

**Full Script Run**
This lists all of the nodes to be updated

```
❯ lando drush scr ../scripts/content/VACMS-3325_paragraph_revision-2020-10.php
Finding Nodes which need to be updated...
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................
Updating Paragraphs
Updating paragraph 3784 on node 255 to use vid 35659
Updating paragraph 3805 on node 255 to use vid 35660
Updating paragraph 3807 on node 255 to use vid 35661
Updating paragraph 3809 on node 255 to use vid 35662
Updating paragraph 5510 on node 255 to use vid 35663
Updating paragraph 5598 on node 78 to use vid 47011
Updating paragraph 2852 on node 201 to use vid 86961
Updating paragraph 9892 on node 284 to use vid 131971
Updating paragraph 7011 on node 284 to use vid 131972
Updating paragraph 4487 on node 720 to use vid 97105
Updating paragraph 4489 on node 720 to use vid 97106
Updating paragraph 4491 on node 720 to use vid 97111
Updating paragraph 4484 on node 720 to use vid 97113
Updating paragraph 3547 on node 67 to use vid 156524
Updating paragraph 11552 on node 6454 to use vid 161405
Updating paragraph 6984 on node 952 to use vid 80491
Updating paragraph 6986 on node 952 to use vid 80492
Updating paragraph 6988 on node 952 to use vid 80499
Updating paragraph 6979 on node 952 to use vid 80501
Updating paragraph 6981 on node 952 to use vid 80503
Updating paragraph 12572 on node 637 to use vid 157706
Updating paragraph 12575 on node 637 to use vid 157709
Updating paragraph 12578 on node 637 to use vid 157712
Updating paragraph 13477 on node 8471 to use vid 161974
Updating paragraph 13478 on node 8471 to use vid 161975
Updating paragraph 13479 on node 8471 to use vid 161976
Updating paragraph 13481 on node 8471 to use vid 161978
Updating paragraph 14016 on node 8471 to use vid 161979
```

## Testing the Patch

To test the patch we save a node in `draft` status which has a previously published revision.  Node 255, 720 and 952 all fit into this category.   After running the data migration script, resave the node while staying in draft status.  Then look at the `paragraphs_item_field_data` for the paragraph revisions to make sure they didn't change.  Using 255 is the easiest since the paragraph ids are already mentioned in this issue.

---
This change is [Reviewable](https://reviewable.io/reviews/department-of-veterans-affairs/va.gov-cms/3409)

### CODE REVIEW

**Comment by ndouglas** [comment]
@indytechcook I ran after 7870df8 but before 5c06c2c and it worked 🎉

**Comment by indytechcook** [comment]
thanks @ndouglas I've found that the patch no longer works.  Can you test and see if saving node `255` updates the `revision_ids` on the `paragraphs_item_field_data` table?

**Comment by indytechcook** [comment]
@ndouglas I updated the patch for the `entity_referece_revisions` module and it does seem to work now.  I must have messed up my testing before.  Please verify.

**Comment by ndouglas** [comment]
After 2839c204, revision IDs are updated correctly.  Anything else you want me to look at?

**Comment by indytechcook** [comment]
@ndouglas just the code itself.

File: scripts/content/VACMS-3325_paragraph_revision-2020-10.php [hunk 1]
```diff
@@ -23,9 +23,9 @@ function run() {
   do {
     getVidsToUpdate($sandbox);
     print('.');
-  } while ($sandbox['current'] / $sandbox['total'] >= 1);
+  } while ($sandbox['total'] - $sandbox['current'] > 0);
```
[the remark this answers is not in the document]
Comment (ndouglas) [reply scripts/content/VACMS-3325_paragraph_revision-2020-10.php:23]: Yeah, it's Monday morning brain.

File: scripts/content/VACMS-3325_paragraph_revision-2020-10.php [hunk 2]
```diff
@@ -108,6 +105,34 @@ function updateNodes($nids) {
...
+
+    $nid = $node->id();
+    $paragraph_field_items = clone $node->get($paragraph->parent_field_name->value);
+    $paragraph_field_items->filter(function($item) use ($paragraph) {
+      return $item->target_id === $paragraph->id();
+    });
+
+    $paragraph_field_item = $paragraph_field_items->first();
+    $vid = $paragraph_field_item->target_revision_id;
+    $vid = $paragraph_field_item->target_revision_id;
```
Comment (ndouglas) [inline scripts/content/VACMS-3325_paragraph_revision-2020-10.php:121]: https://www.youtube.com/watch?v=jGVi_zrH38o

**Review by ndouglas — COMMENTED** [review]
I don't think I have any functional criticisms, just a doubled LOC and a function that doesn't appear to be called anywhere.

**Review by ndouglas — APPROVED** [review]

File: scripts/content/VACMS-3325_paragraph_revision-2020-10.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    $sandbox['pids'] += $new_pids;
+  }
+
+  $sandbox['current'] += count($paragraph_ids);
+}
+
+function runRange($paragraph_ids) : array {
+  $pids = [];
+  try {
+    $paragraphs = getParagraphData($paragraph_ids);
```
Comment (acrollet) [inline scripts/content/VACMS-3325_paragraph_revision-2020-10.php:45]: if I was wearing my new clean code hat I'd say that a name like getLoadedParagraphs might be more representative of what's happening. :-) Don't see a reason to make a change for a one-time script though.

Comment (indytechcook) [reply scripts/content/VACMS-3325_paragraph_revision-2020-10.php:45]: 💯

File: scripts/content/VACMS-3325_paragraph_revision-2020-10.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    return $item->target_id === $paragraph->id() && $paragraph->getRevisionId() !== $item->target_revision_id;
+  });
+
+  return count($paragraph_field_items) > 0;
+}
+
+function updateParagraphs($paragraphs) {
+  /** @var \Drupal\paragraphs\Entity\Paragraph $paragraph */
+  foreach ($paragraphs as $paragraph) {
+    $node = $paragraph->getParentEntity();
```
Comment (acrollet) [inline scripts/content/VACMS-3325_paragraph_revision-2020-10.php:101]: nitpicky but do we have a guarantee that this is a node? Ah, perhaps we do from shouldWeUpdateThisParagraph(). carry on!

**Review by acrollet — APPROVED** [review]

**Comment by acrollet** [comment]
Documenting that the script worked for me as of 074739f:

before script:
```
MySQL [drupal8]> SELECT id, revision_id
    -> FROM paragraphs_item 
    -> WHERE id in (3784, 3805, 3807, 3809, 5510);
+------+-------------+
| id   | revision_id |
+------+-------------+
| 3784 |       37275 |
| 3805 |       37276 |
| 3807 |       37277 |
| 3809 |       37278 |
| 5510 |       37279 |
+------+-------------+
5 rows in set (0.01 sec)

MySQL [drupal8]> select field_content_block_target_id, field_content_block_target_revision_id from node__field_content_block where entity_id=255;
+-------------------------------+----------------------------------------+
| field_content_block_target_id | field_content_block_target_revision_id |
+-------------------------------+----------------------------------------+
|                          3784 |                                  35659 |
|                          3805 |                                  35660 |
|                          3807 |                                  35661 |
|                          3809 |                                  35662 |
|                          5510 |                                  35663 |
+-------------------------------+----------------------------------------+
5 rows in set (0.01 sec)
```

Script:
```Finding Nodes which need to be updated...
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................
Updating Paragraphs
Updating paragraph 3784 on node 255 to use vid 35659
Updating paragraph 3805 on node 255 to use vid 35660
Updating paragraph 3807 on node 255 to use vid 35661
Updating paragraph 3809 on node 255 to use vid 35662
Updating paragraph 5510 on node 255 to use vid 35663
Updating paragraph 5598 on node 78 to use vid 47011
Updating paragraph 2852 on node 201 to use vid 86961
Updating paragraph 9892 on node 284 to use vid 131971
Updating paragraph 7011 on node 284 to use vid 131972
Updating paragraph 4487 on node 720 to use vid 97105
Updating paragraph 4489 on node 720 to use vid 97106
Updating paragraph 4491 on node 720 to use vid 97111
Updating paragraph 4484 on node 720 to use vid 97113
Updating paragraph 3547 on node 67 to use vid 156524
Updating paragraph 11552 on node 6454 to use vid 161405
Updating paragraph 6984 on node 952 to use vid 80491
Updating paragraph 6986 on node 952 to use vid 80492
Updating paragraph 6988 on node 952 to use vid 80499
Updating paragraph 6979 on node 952 to use vid 80501
Updating paragraph 6981 on node 952 to use vid 80503
Updating paragraph 12572 on node 637 to use vid 157706
Updating paragraph 12575 on node 637 to use vid 157709
Updating paragraph 12578 on node 637 to use vid 157712
Updating paragraph 13477 on node 8471 to use vid 161974
Updating paragraph 13478 on node 8471 to use vid 161975
Updating paragraph 13479 on node 8471 to use vid 161976
Updating paragraph 13481 on node 8471 to use vid 161978
Updating paragraph 14016 on node 8471 to use vid 161979
```

After script:
```
MySQL [drupal8]> select field_content_block_target_id, field_content_block_target_revision_id from node__field_content_block where entity_id=255;
+-------------------------------+----------------------------------------+
| field_content_block_target_id | field_content_block_target_revision_id |
+-------------------------------+----------------------------------------+
|                          3784 |                                  35659 |
|                          3805 |                                  35660 |
|                          3807 |                                  35661 |
|                          3809 |                                  35662 |
|                          5510 |                                  35663 |
+-------------------------------+----------------------------------------+
5 rows in set (0.00 sec)

MySQL [drupal8]> SELECT id, revision_id FROM paragraphs_item  WHERE id in (3784, 3805, 3807, 3809, 5510);
+------+-------------+
| id   | revision_id |
+------+-------------+
| 3784 |       35659 |
| 3805 |       35660 |
| 3807 |       35661 |
| 3809 |       35662 |
| 5510 |       35663 |
+------+-------------+
5 rows in set (0.00 sec)
```

**Pull request merged**